### PR TITLE
Make AppDataPath absolute against the AppWorkPath if it is not

### PR DESCRIPTION
**Pull request by zeripath**
There are multiple repeated issues whereby a non-absolute provided
APP_DATA_PATH causes strange issues.

This PR simply absolutes the APP_DATA_PATH against the AppWorkPath if
its not so. It also ensures that AppWorkPath is also always absolute.

Ref #19367

Signed-off-by: Andrew Thornton <art27@cantab.net>

## :warning: BREAKING :warning: 

Server administrators should be aware that both the work path and the data path are now made absolute.

* The work path will be made absolute against the PWD and if that is not possible against the AppPath
* The APP_DATA_PATH will be made absolute against the work path. 
 
This second change is a breaking change for administrators who had APP_DATA_PATH as a relative path and the work path was not the current PWD.

However, because of the multiple issues reported with non-absolute APP_DATA_PATHs we expect that this change will affect few people.

### CODE REVIEW

**Comment by techknowlogick** [comment]
Should we log a warn or info now we are ensuring absolute paths against workpath? Either way this LGTM

**Comment by silentcodeg** [comment]
This is a breaking change for all existing Gitea installations that have a relative AppDataPath in app.ini that is not against AppWorkPath.

File: modules/setting/setting.go [hunk 1]
```diff
@@ -478,6 +478,10 @@ func getWorkPath(appPath string) string {
 			workPath = appPath[:i]
 		}
 	}
+	workPath = strings.ReplaceAll(workPath, "\\", "/")
```
Comment (silverwind) [inline modules/setting/setting.go:481]: `filepath.Join` can deal with backslashes on Windows. It doesn't need such hacks. After all, backslash is a valid path character on many file sytems.

Comment (zeripath) [reply modules/setting/setting.go:481]: This is to be consistent with the rest of the filepath munging in this file.

**Comment by zeripath** [comment]
> Should we log a warn or info now we are ensuring absolute paths against workpath? Either way this LGTM

Have done so.

**Comment by zeripath** [comment]
> This is a breaking change for all existing Gitea installations that have a relative AppDataPath in app.ini that is not against AppWorkPath.

I'm not sure that such configurations previously worked properly in any case. 

I can add a breaking notice.

**Review by lunny — APPROVED** [review]

**Pull request merged**